### ISSUE TITLE
icdf not implemented in Truncated Distribution

### ISSUE THREAD

**Issue by TheSkyentist** [pyro-ppl/numpyro#1937]
**Context:** 
I am attempted to perform Nested Sampling using the contributed `jaxns` wrapper. Part of what the wrapper attempts to do (I believe) is re-parametrize my priors to Uniform distributions (which makes sense). However this fails as the `icdf` method is not implemented for Truncated Distributions which best reflect (some of) my priors. 

**Minimal Work Example:**
```python
#! /usr/bin/env python

# JAX packages
from jax import random
from jax import numpy as jnp

# Bayesian Inference
from numpyro import distributions as dist, sample
from numpyro.contrib.nested_sampling import NestedSampler

# Random key
rng_key = random.PRNGKey(0)

# Define data
x, err = 0.0, 1 / jnp.sqrt(2 * jnp.pi)

def model(x, err, π):
    center = sample('center', dist.TruncatedNormal(0, π, low = -10, high = 10))

    # Delta function likelihood
    sample('obs', dist.Normal(center, err), obs=x)


# Model Arguements
model_args = (x, err, 5)

# Nested Sampling
constructor_kwargs = {'num_live_points': 1000, 'max_samples': 50000}
termination_kwargs = {'dlogZ': 0.01}
NS = NestedSampler(
    model=model,
    constructor_kwargs=constructor_kwargs,
    termination_kwargs=termination_kwargs,
)
NS.run(rng_key, *model_args)
NS.print_summary()
```
This fails with a `NotImplementedError` as `icdf`s are not implemented for Truncated distributions. 

**Steps Forward?**
Would it be easier to implement the `icdf` method for truncated distributions or perhaps is there a better way to transform my priors such that they are not exposed to this shortcoming? If the former is preferred, I'm happy to try and contribute to the effort, I wonder if there are special considerations that would need to be taken into account to maintain the correct statistics.

**Comment by TheSkyentist** [comment]
The documentation suggests that the efficient way for building Truncated Distributions in the first place is to calculate their ICDF:
https://num.pyro.ai/en/stable/tutorials/truncated_distributions.html#3.-Sampling-from-truncated-and-folded-distributions

Whether this is how it is implemented currently remains to be seen, but perhaps allowing this implementation is just a matter of exposing existing functionality.

**Comment by martinjankowiak** [comment]
implementing icdf for truncatednormal should be straightforward, although these kinds of functions can potentially be numerically inaccurate/unstable depending on the args, so it behooves some testing. https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.erf_inv.html

it would be trickier though to implement icdf for all truncated distributions, since in some cases there is no closed-form icdf or the icdf involves nasty special functions which may not be available in jax.

**Comment by tillahoffmann** [comment]
I think the [`sample`](https://github.com/pyro-ppl/numpyro/blob/4b33db15576b8f20af6d13772ad13100f877f16a/numpyro/distributions/truncated.py#L232-L250) method already implements the icdf of the truncated distribution. It might just be a matter of moving some of that code to the `icdf` method and calling from `sample` to avoid redundancy.

**Comment by TheSkyentist** [comment]
@tillahoffmann My thoughts exactly!

**Comment by TheSkyentist** [comment]
I've started a draft PR (#1938). I need to run tests, but hopefully it's as simple as that.

**Comment by TheSkyentist** [comment]
Testing under #1938 does indeed resolve the issue with `jaxns`.

**Comment by TheSkyentist** [comment]
Fixed with the merge of #1938
